src/validationinterface.cpp

```diff
@@ -16,36 +16,59 @@
-    boost::signals2::scoped_connection UpdatedBlockTip;
-    boost::signals2::scoped_connection TransactionAddedToMempool;
-    boost::signals2::scoped_connection BlockConnected;
-    boost::signals2::scoped_connection BlockDisconnected;
-    boost::signals2::scoped_connection TransactionRemovedFromMempool;
-    boost::signals2::scoped_connection ChainStateFlushed;
-    boost::signals2::scoped_connection BlockChecked;
-    boost::signals2::scoped_connection NewPoWValidBlock;
-};
-
 struct MainSignalsInstance {
-    boost::signals2::signal<void (const CBlockIndex *, const CBlockIndex *, bool fInitialDownload)> UpdatedBlockTip;
-    boost::signals2::signal<void (const CTransactionRef &)> TransactionAddedToMempool;
-    boost::signals2::signal<void (const std::shared_ptr<const CBlock> &, const CBlockIndex *pindex)> BlockConnected;
-    boost::signals2::signal<void (const std::shared_ptr<const CBlock>&, const CBlockIndex* pindex)> BlockDisconnected;
-    boost::signals2::signal<void (const CTransactionRef &)> TransactionRemovedFromMempool;
-    boost::signals2::signal<void (const CBlockLocator &)> ChainStateFlushed;
-    boost::signals2::signal<void (const CBlock&, const BlockValidationState&)> BlockChecked;
-    boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
-
     // We are not allowed to assume the scheduler only runs in one thread,
     // but must ensure all callbacks happen in-order, so we end up creating
     // our own queue here :(
     SingleThreadedSchedulerClient m_schedulerClient;
-    std::unordered_map<CValidationInterface*, ValidationInterfaceConnections> m_connMainSignals;
+    Mutex m_mutex;
+    struct Entry { std::shared_ptr<CValidationInterface> callbacks; int count = 1; };
+    std::list<Entry> m_list GUARDED_BY(m_mutex);
+    std::unordered_map<CValidationInterface*, std::list<Entry>::iterator> m_map GUARDED_BY(m_mutex);
 
     explicit MainSignalsInstance(CScheduler *pscheduler) : m_schedulerClient(pscheduler) {}
+
+    void Add(std::shared_ptr<CValidationInterface> callbacks)
+    {
+        LOCK(m_mutex);
+        auto inserted = m_map.emplace(callbacks.get(), m_list.end());
+        if (inserted.second) {
+            m_list.emplace_back();
+            inserted.first->second = std::prev(m_list.end());
+        }
+        inserted.first->second->callbacks = std::move(callbacks);
+    }
+
+    void Remove(CValidationInterface* key)
+    {
+        LOCK(m_mutex);
+        auto it = m_map.find(key);
+        if (it != m_map.end()) {
+            if (--it->second->count == 0) m_list.erase(it->second);
+            m_map.erase(it);
+        }
+    }
+
+    void Clear()
+    {
+        LOCK(m_mutex);
+        for (auto it = m_list.begin(); it != m_list.end();) {
+            if (--it->count == 0) it = m_list.erase(it); else ++it;
+        }
+        m_map.clear();
+    }
+
+    template<typename F> void Iterate(F&& f)
+    {
+        WAIT_LOCK(m_mutex, lock);
```

Comment:
How about local copy and then iterate it lock free?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18524#discussion_r403493569

> How about local copy and then iterate it lock free?

I'm not sure what approach could be entirely lock free since the list is global and can be modified from any thread.

I just implemented something simple that can be changed and optimized in the future. The lock is not held when calling CValidation interface methods. It avoids copying so allocations aren't needed just to iterate the list. Assumption is that the list is iterated frequently and modified infrequently

Reply:
> Assumption is that the list is iterated frequently and modified infrequently

Right. What I had in mind is 2 lists and an atomic book `list_changed`. On iterating it would then sync the list with the mutex locked.

Reply:
I believe what @promag suggests is: grab lock, copy the entire list, release lock, and the execute. Which is not lock free, but has no locks during execution.

As I expect that the vast majority of the locks here will be uncontested, I think the current approach is better (uncontested lock grabs are faster than mallocs, I think).

Reply:
So at best case it would be lock free I think.

Reply:
Oops, I wanted to delete my own comment, but it seems I misclicked and deleted @promag's. Restoring from mail:

> > Assumption is that the list is iterated frequently and modified infrequently
>
> Right. What I had in mind is 2 lists and an atomic book list_changed. On iterating it would then sync the list with the mutex locked.

Reply:
:) I was wondering what happened. So draf:
```
if (list_changed) {
  lock;
  iterate_list = list;
  list_changed = false
}
for (f : iterate_list) f()
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18524#discussion_r403493569

> Right. What I had in mind is 2 lists and an atomic book list_changed. On iterating it would then sync the list with the mutex locked.

I don't understand the suggestion from the pseudocode. I don't love the idea of having multiple lists, but even if you have them, I don't see how you avoid locks copying the list if you do create copies, or avoid needing a condition variable to delay modifying the list if you don't create copies. I'm probably just making an incorrect assumption about what you want to accomplish here. Feel free to post sample code, or just change the implementation in an alternate PR or followup PR. I'd be especially interested if it's a simplification and not just an optimization.

Reply:
Is it not possible that there are two invocations of Iterate simultaneously?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18524#discussion_r403600366

> Is it not possible that there are two invocations of Iterate simultaneously?

It shouldn't be the most common thing but should be possible because notifications like CMainSignals::BlockChecked and CMainSignals::NewPoWValidBlock run on calling thread. Other notifications are sent from the scheduler so shouldn't happen simultaneously.

Reply:
> Is it not possible that there are two invocations of Iterate simultaneously?

My previous suggestion isn't possible in this case.